src/Build/BackEnd/Components/SdkResolution/MainNodeSdkResolverService.cs

```diff
@@ -78,148 +57,57 @@ public override void ClearCaches()
-                        response = ResolveSdk(request.SubmissionId, sdkReference, new EvaluationLoggingContext(loggingService, request.BuildEventContext, request.ProjectPath), request.ElementLocation, request.SolutionPath, request.ProjectPath, request.Interactive, request.IsRunningInVisualStudio);
-                    }
-                    catch (Exception e)
-                    {
-                        ILoggingService loggingService = Host.GetComponent(BuildComponentType.LoggingService) as ILoggingService;
-
-                        EvaluationLoggingContext loggingContext = new EvaluationLoggingContext(loggingService, request.BuildEventContext, request.ProjectPath);
-
-                        loggingService.LogFatalBuildError(loggingContext.BuildEventContext, e, new BuildEventFileInfo(request.ElementLocation));
-                    }
-                    finally
-                    {
-                        // Get the node manager and send the response back to the node that requested the SDK
-                        INodeManager nodeManager = Host.GetComponent(BuildComponentType.NodeManager) as INodeManager;
-
-                        nodeManager.SendData(request.NodeId, response);
-                    }
-                }));
-            }
-
-            // Wait for all tasks to complete
-            Task.WaitAll(tasks.ToArray());
-        }
+                    ILoggingService loggingService = Host.GetComponent(BuildComponentType.LoggingService) as ILoggingService;
 
-        /// <summary>
-        /// A background thread that waits for requests to be received.
-        /// </summary>
-        private void RequestHandlerPumpProc()
-        {
-            try
-            {
-                Thread.CurrentThread.Name = "MSBuild SDK Resolver";
+                    EvaluationLoggingContext loggingContext = new EvaluationLoggingContext(loggingService, request.BuildEventContext, request.ProjectPath);
 
-                while (true)
+                    loggingService.LogFatalBuildError(loggingContext.BuildEventContext, e, new BuildEventFileInfo(request.ElementLocation));
+                }
+                finally
                 {
-                    WaitHandle[] handles = new WaitHandle[] { ShutdownEvent, _requestReceivedEvent };
-
-                    int waitId = WaitHandle.WaitAny(handles);
-                    switch (waitId)
-                    {
-                        case 0:
-                            return;
+                    // Get the node manager and send the response back to the node that requested the SDK
+                    INodeManager nodeManager = Host.GetComponent(BuildComponentType.NodeManager) as INodeManager;
 
-                        case 1:
-                            _requestReceivedEvent.Reset();
+                    nodeManager.SendData(request.NodeId, response);
+                }
+            }).ConfigureAwait(continueOnCapturedContext: false);
```

Comment:
nit: Does `continueOnCapturedContext: false` make a difference when the task has no continuation?

Reply:
The more I look into it, it seems like this method is already running on a thread pool thread.  The communication receives packets and runs these with a worker queue.  Maybe I don't even need to use `Task.Run()` at all.  Let me look into it.

Reply:
Latest version doesn't call `Task.Run()` this this method is already running in a thread pool

Reply:
Wait, it is running in the thread pool but it doesn't mean that it can be synchronous. Here's the callstack I see with my test build:

```
 	Microsoft.Build.dll!Microsoft.Build.BackEnd.SdkResolution.MainNodeSdkResolverService.PacketReceived(int node, Microsoft.Build.BackEnd.INodePacket packet) Line 81	C#
 	Microsoft.Build.dll!Microsoft.Build.BackEnd.NodePacketFactory.PacketFactoryRecord.RoutePacket(int nodeId, Microsoft.Build.BackEnd.INodePacket packet) Line 110	C#
 	Microsoft.Build.dll!Microsoft.Build.BackEnd.NodePacketFactory.PacketFactoryRecord.DeserializeAndRoutePacket(int nodeId, Microsoft.Build.BackEnd.ITranslator translator) Line 102	C#
 	Microsoft.Build.dll!Microsoft.Build.BackEnd.NodePacketFactory.DeserializeAndRoutePacket(int nodeId, Microsoft.Build.BackEnd.NodePacketType packetType, Microsoft.Build.BackEnd.ITranslator translator) Line 58	C#
 	Microsoft.Build.dll!Microsoft.Build.BackEnd.NodeManager.DeserializeAndRoutePacket(int nodeId, Microsoft.Build.BackEnd.NodePacketType packetType, Microsoft.Build.BackEnd.ITranslator translator) Line 261	C#
 	Microsoft.Build.dll!Microsoft.Build.BackEnd.NodeProviderOutOfProcBase.NodeContext.ReadAndRoutePacket(Microsoft.Build.BackEnd.NodePacketType packetType, byte[] packetData, int packetLength) Line 1090	C#
>	Microsoft.Build.dll!Microsoft.Build.BackEnd.NodeProviderOutOfProcBase.NodeContext.BodyReadComplete(System.IAsyncResult result) Line 1144	C#
```

The frame at the bottom has:
```C#
                    // Read the next packet.
                    BeginAsyncPacketRead();
```
as the next statement to execute. Can we really afford to block processing of further packets from the node until the SDK is resolved?

Reply:
I was looked at it. With current state of code we can afford to block packet processing until SDK is resolved. Nodes are single threaded, and node is blocked and wait for response while it sends request to main node for SDK resolution. So it is guaranteed, if I understand it correctly, that there is actually no other packet to be processed - on that particular namedpipe.